### PR TITLE
Lua wesnoth.deprecate_api bug when deprecating tables with metatables

### LINKED ISSUE #4079: Status icons not showing in sidebar in Dead Water
When a unit is poisoned or slowed in 1.15.0-dev, the status icons don't show for the unit in the sidebar.

### PR DESCRIPTION
**Pull request by jostephd**
See commit messages for details.

Fixes #4079 

Test case: start the test scenario, click on the petrified Elvish Avenger, and see whether the "petrified" icon appears in the sidebar.

### CODE REVIEW

File: data/lua/core.lua [hunk 1]
```diff
@@ -202,26 +202,32 @@ function wesnoth.deprecate_api(elem_name, replacement, level, version, elem, det
...
-					wesnoth.deprecated_message(elem_name, level, version, message)
-				end
-				elem[key] = val
-			end,
-		}
 		-- Don't clobber the old metatable.
 		local old_mt = getmetatable(elem)
-		if type(old_mt) == "table" then
-			setmetatable(mt, old_mt)
+		if old_mt == nil then old_mt = {} end
```
Comment (CelticMinstrel) [inline data/lua/core.lua]: This line could be merged with the previous one as `local old_mt = getmetatable(elem) or {}`.

File: data/lua/core.lua [hunk 2]
```diff
@@ -202,26 +202,32 @@ function wesnoth.deprecate_api(elem_name, replacement, level, version, elem, det
...
+		local mt = {}
+		for k,v in pairs(old_mt) do
+			mt[k] = old_mt[v]
+		end
+		mt.__index = function(self, key)
+			if not msg_shown then
+				msg_shown = true
+				wesnoth.deprecated_message(elem_name, level, version, message)
+			end
+			if type(old_mt) == "table" and old_mt.__index ~= nil then
```
Comment (CelticMinstrel) [inline data/lua/core.lua]: You also need to handle the case where `type(old_mt.__index) == 'table'`.

(Just for the record, that's only possible for `__index`, not for `__newindex`.)

File: data/lua/core.lua (hunk 1 above)
Comment (jostephd) [reply data/lua/core.lua]: Fixed

File: data/lua/core.lua (hunk 2 above)
Comment (jostephd) [reply data/lua/core.lua]: Thanks, I missed that in the lua docs. I forced push a fix. I wanted to add an `else` case that would log an error message but I couldn't get `error('foo')` to print an error message.

Comment (CelticMinstrel) [reply data/lua/core.lua]: Maybe try `wml.error()` instead (formerly know as `helper.wml_error()`)? I guess it's not the ideal thing here though...

**Comment by jostephd** [comment]
Thanks for the review!

> Maybe try wml.error() instead (formerly know as helper.wml_error())? I guess it's not the ideal thing here though...

That worked, the error is reported in the chat. It's better than nothing so i pushed that and will merge.

**Pull request merged**